Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_delegate.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class ListenSequence;
+class MemoryPersistence;
+class Sizer;
+
+/**
+ * Provides the LRU GC implementation for memory persistence.
+ */
+class MemoryLruReferenceDelegate : public LruDelegate {
+ public:
+  MemoryLruReferenceDelegate(MemoryPersistence* persistence,
+                             LruParams lruParams,
+                             std::unique_ptr<Sizer> sizer);
+
+  bool IsPinnedAtSequenceNumber(model::ListenSequenceNumber upperBound,
+                                const model::DocumentKey& key);
+
+  // MARK: ReferenceDelegate overrides
+
+  model::ListenSequenceNumber current_sequence_number() override;
+
+  void AddInMemoryPins(ReferenceSet* set) override;
+
```

Comment:
Can some of these functions be grouped together? E.g.,
```cpp
void AddInMemoryPins(ReferenceSet* set) override;
void AddReference(const model::DocumentKey& key) override;

void RemoveReference(const model::DocumentKey& key) override;
void RemoveMutationReference(const model::DocumentKey& key) override;
void RemoveTarget(const QueryData& queryData) override;

void UpdateLimboDocument(const model::DocumentKey& key) override;

void OnTransactionStarted(absl::string_view label) override;
void OnTransactionCommitted() override;
```
I find the current layout hard to read.

Reply:
I'm fine with using a different criteria for grouping, e.g.:
```cpp
void AddInMemoryPins(ReferenceSet* set) override;

void AddReference(const model::DocumentKey& key) override;
void RemoveReference(const model::DocumentKey& key) override;

void RemoveMutationReference(const model::DocumentKey& key) override;

void RemoveTarget(const QueryData& queryData) override;

void UpdateLimboDocument(const model::DocumentKey& key) override;

void OnTransactionStarted(absl::string_view label) override;
void OnTransactionCommitted() override;
```
I personally don't interpret that if `RemoveReference` and `RemoveMutationReference` are next to each other, this means they're very closely related -- more like "This class can be used to remove things, here's a list of what can be removed". On the other hand, I can see how this might be confusing. With that said, I'd argue in favor of _some_ grouping; it can't be the case that a class is a set of entirely unrelated methods, which is what a lack of grouping seems to imply.

Reply:
The lack of grouping flows from the parent that specifies the interface. In the parent the methods are ungrouped and seemingly rightly so, given that each needs a distinct comment and has a different behavior. It seems strange to me to then arbitrarily apply a different grouping in the subclass that's implementing the interface.

I suppose one alternative would be to group all of these methods together into one clump, essentially saying that these `override` declarations together exist to indicate that indeed this does implement that interface, but are otherwise uninteresting.

Reply:
My personal preference is to always separate methods if they have comments (or sometimes if the declaration spans multiple lines), but to group them if there are no comments. Given they're all overrides, it sounds reasonable to just lump them all together. But if you feel strongly they should be ungrouped, I'm okay with that.

Reply:
OK. I don't care strongly enough about this to fight about it. I've added the grouping.